docs/docs/09-rfcs/2023-03-08-wing-preview-environment.md

```diff
@@ -0,0 +1,278 @@
+
+Every possible interaction with the Wing Cloud will be through an API. Any request to the API will be authenticated using this cookie. The API will then use the user ID to check if the user has the necessary permissions to perform the action.
+
+Since we want to go minimal, we can use the GitHub API and the user permissions to the repository as the authorization mechanism. Beware the GitHub API has a [rate limit](https://docs.github.com/en/apps/creating-github-apps/registering-a-github-app/rate-limits-for-github-apps#installation-access-tokens-on-githubcom), so we may want to cache the permissions for a while if we foresee hitting these limits. Also, see the [List collaborators endpoint](https://docs.github.com/en/free-pro-team@latest/rest/collaborators/collaborators?apiVersion=2022-11-28#list-repository-collaborators) for more information on authorization.
+
+#### Process: Processing GitHub Events
+
+A Cloud.Function will listen to GitHub events and will process them. Based on the event, the function will create, update or delete preview environments on Flyio.
+
+```mermaid
+graph LR
+    gh{GitHub}
+    ghapp>GitHub App]
+    fn[Process Events Fn]
+    flyio{Flyio}
+    gh -- Integration --> ghapp
+    ghapp -- Events --> fn
+    fn -- Create Preview Environment --> flyio
+    fn -- Update Preview Environment --> flyio
+    fn -- Delete Preview Environment --> flyio
+```
+
+Depending on whether GitHub retries events or not, we may want to include a FIFO queue in the middle.
+
+```mermaid
+graph LR
+    gh{GitHub}
+    ghapp>GitHub App]
+    fn[Process Events Fn]
+    queue[(Events Queue)]
+    fn2[Enqueue Events Fn]
+    ghapp -- Events --> fn2
+    fn2 --> queue
+    queue --> fn
+    gh -- Integration --> ghapp
+    fn --> rest[...]
+```
+
+The probot npm package will take care of ensuring the events received are legitimate.
+
+#### Process: Creating Preview Environments
+
+When a GitHub event `pull_request.opened` is received, we will look for `*.main.w` files and create a preview environment for each of them. We will use the Flyio API to create a new project(s) and deploy the code to it (the Flyio specifics may change).
+
+New subdomains will be created for each preview environment:
+
+- `<gh-repository>-<nanoid>.wingcloud.app`
+- `<gh-repository>-<gh-branch>-<entry-point>.wingcloud.app`, as an alias of the above, if available. Otherwise, a random ID could be added somewhere, such as in `<gh-repository>-<gh-branch>-<nanoid>-<entry-point>.wingcloud.app`
+
+During this process, we will also create new entries in the Cloud.Table to store the Flyio project ID, the GitHub PR ID, the entry points and whatever unique IDs we generate for every environment. This data is necessary to later resolve the subdomains to the Flyio projects, to update them later, and to delete them.
+
+The technical details on how the subdomain resolution will work at DNS level are still to be determined.
```

Comment:
Did we ever discuss a `cloud.Dns` abstraction? Wouldn't this be useful here?

Reply:
I don't recall, but sure it'd be helpful.